internal/metrics/metrics.go

```diff
@@ -0,0 +1,199 @@
+	AddMetric(value float64, labelsMap map[string]string) error
+}
+
+type TSDB struct {
+	appliedOptions *tsdb.Options
+	db             *tsdb.DB
+	dbLock         sync.RWMutex
+}
+
+func NewTSDB(dataDir string) (*TSDB, error) {
+	metricsDir := path.Join(dataDir, "metrics")
+	if err := os.MkdirAll(metricsDir, 0755); err != nil {
+		return nil, fmt.Errorf("cannot create directory: %w", err)
+	}
+	opts := tsdb.DefaultOptions()
+
+	opts.RetentionDuration = int64(DefaultRetentionDuration / time.Millisecond)
+	opts.MaxBytes = DefaultMaxBytes
+
+	db, err := tsdb.Open(metricsDir, nil, nil, opts, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &TSDB{
+		db:             db,
+		dbLock:         sync.RWMutex{},
+		appliedOptions: opts,
+	}, err
+}
+
+func (t *TSDB) Deregister() error {
+	metricsDir := t.db.Dir()
+	t.dbLock.Lock()
+	defer t.dbLock.Unlock()
+
+	err := t.db.Close()
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	err = os.RemoveAll(metricsDir)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	return nil
+}
+
+func (t *TSDB) GetMetricsForTimeRange(tMin time.Time, tMax time.Time) ([]Series, error) {
+	log.Tracef("Getting metrics for %v - %v", tMin, tMax)
+	q, err := t.getDB().Querier(nil, toDbTime(tMin), toDbTime(tMax))
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		err = q.Close()
+		if err != nil {
+			log.Error(err)
+		}
+	}()
+
+	matchers, err := getAllLabelsMatchers(q)
+	if err != nil {
+		return nil, err
+	}
+	seriesSet := q.Select(false, nil, matchers...)
+	if seriesSet.Err() != nil {
+		log.Error(seriesSet.Err())
+		return nil, seriesSet.Err()
+	}
+
+	return toSeries(seriesSet), nil
+}
+
+func toSeries(seriesSet storage.SeriesSet) []Series {
+	var allSeries []Series
+	for {
+		if !seriesSet.Next() {
+			break
+		}
+		s := seriesSet.At()
+		labelsMap := make(map[string]string)
+		for _, label := range s.Labels() {
+			labelsMap[label.Name] = label.Value
+		}
+		var dataPoints []DataPoint
+		for i := s.Iterator(); i.Next(); {
+			t, v := i.At()
+			dataPoints = append(dataPoints, DataPoint{Time: t, Value: v})
+		}
+		allSeries = append(allSeries, Series{
+			Labels:     labelsMap,
+			DataPoints: dataPoints,
+		})
+	}
+	return allSeries
+}
+
+func (t *TSDB) AddMetric(value float64, labelsMap map[string]string) error {
+	appender := t.getDB().Appender(context.Background())
+	_, err := appender.Append(0, labels.FromMap(labelsMap), toDbTime(time.Now()), value)
+	if err != nil {
+		return err
+	}
+
+	return appender.Commit()
+}
+
+func (t *TSDB) Update(config models.DeviceConfigurationMessage) error {
+	metricsConfiguration := config.Configuration.Metrics
+	maxBytes := DefaultMaxBytes
+	retentionDuration := DefaultRetentionDuration / time.Millisecond
+	if metricsConfiguration != nil && metricsConfiguration.Retention != nil {
+		maxBytes = int64(metricsConfiguration.Retention.MaxMib) * 1024 * 1024
+		retentionDuration = time.Duration(metricsConfiguration.Retention.MaxHours) * time.Hour / time.Millisecond
```

Comment:
I don't think it is OK to set the duration directly with a number. Although Duration is an int64 it is supposed to be in nanoseconds and we set it to hours. We then use the fact that it is an int64 for converting to nanoseconds and then back to milliseconds.
I think we should not use the time package in this case.

Reply:
You're right - I will change it.

Reply:
Done.